go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -959,16 +958,30 @@ func (npw *nodePortWatcher) UpdateEndpointSlice(oldEpSlice, newEpSlice *discover
 
 	// Update rules and service cache if hasHostNetworkEndpoints status changed or localEndpoints changed
 	nodeIPs := npw.nodeIPManager.ListAddresses()
-	hasLocalHostNetworkEpOld := hasLocalHostNetworkEndpoints([]*discovery.EndpointSlice{oldEpSlice}, nodeIPs, svc)
-	hasLocalHostNetworkEpNew := hasLocalHostNetworkEndpoints([]*discovery.EndpointSlice{newEpSlice}, nodeIPs, svc)
 	epSlices, err := npw.watchFactory.GetEndpointSlices(newEpSlice.Namespace, newEpSlice.Labels[discovery.LabelServiceName])
 	if err != nil {
-		klog.V(5).Infof("Could not fetch endpointslices for service %s during endpointSliceDelete", namespacedName)
+		if !kerrors.IsNotFound(err) {
+			return fmt.Errorf("error retrieving all endpointslices for service %s/%s during endpointslice update on %s: %w",
+				namespacedName.Namespace, namespacedName.Name, newEpSlice.Name, err)
+		}
+		klog.V(5).Infof("No endpointslices found for service %s/%s during endpointslice update on %s: %v",
+			namespacedName.Namespace, namespacedName.Name, newEpSlice.Name, err)
 	}
 
+	// Delete old endpoint slice and add new one when local endpoints have changed or the presence of local host-network
+	// endpoints has changed. For this second comparison, check first between the old endpoint slice and all current
+	// endpointslices for this service. This is a partial comparison, in case serviceInfo is not set. When it is set, compare
+	// between /all/ old endpoint slices and all new ones.
+	localEndpointsFromOldEndpointSlice := npw.GetLocalEndpointAddresses([]*discovery.EndpointSlice{oldEpSlice}, svc)
```

Comment:
I changed it on purpose to draw the attention on the fact that `newLocalEndpoints` is created from all endpointslices for this service, whereas `localEndpointsFromOldEndpointSlice` is created only from the old endpoint slice we got an update from. But I added a comment that explains it already, so I can name the variable `oldLocalEndpoints` in the end... :)

Reply:
yeah I see what you mean,, I admit I missed your detailed comment above it on the partial versus full comparison, I guess the perf gain is considerably higher than the comparison overhead -> probably in larges clusters this will be a huge gain but in smaller clusters with like 10 endpoints this would not matter much i guess.
but lgtm on this for now.